src/shell/commands/disk_rebalance.cpp

```diff
@@ -0,0 +1,293 @@
+        }
+
+        if (flags.find(flag) == flags.end()) {
+            fmt::print(stderr, "unknown flag {}\n", flag);
+            return false;
+        }
+    }
+
+    return true;
+}
+
+std::ostream *get_out_stream(const std::string &file_name)
+{
+    std::ostream *os_ptr = nullptr;
+    if (file_name.empty()) {
+        os_ptr = &std::cout;
+    } else {
+        os_ptr = new std::ofstream(file_name);
+        if (!*os_ptr) {
+            fmt::print(stderr, "open output file {} failed\n", file_name);
+            delete os_ptr;
+        }
+    }
+    return os_ptr;
+}
+
+bool query_disk_info(
+    shell_context *sc,
+    const argh::parser &cmd,
+    const std::string &node_address,
+    const std::string &app_name,
+    /*out*/ std::map<dsn::rpc_address, dsn::error_with<query_disk_info_response>> &err_resps)
+{
+    std::map<dsn::rpc_address, dsn::replication::node_status::type> nodes;
+    auto error = sc->ddl_client->list_nodes(::dsn::replication::node_status::NS_INVALID, nodes);
+    if (error != dsn::ERR_OK) {
+        fmt::print(stderr, "list nodes failed, error={}\n", error.to_string());
+        return false;
+    }
+
+    std::vector<dsn::rpc_address> targets;
+    if (!node_address.empty()) {
+        for (const auto &node : nodes) {
+            if (node.first.to_std_string() == node_address) {
+                targets.push_back(node.first);
+                break;
+            }
+        }
+    } else {
+        for (const auto &node : nodes) {
+            targets.push_back(node.first);
+        }
+    }
+
+    if (targets.empty()) {
+        fmt::print(stderr, "invalid target replica server address!\n");
+        return false;
+    }
+    sc->ddl_client->query_disk_info(targets, app_name, err_resps);
+    return true;
+}
+
+bool query_disk_capacity(command_executor *e, shell_context *sc, arguments args)
+{
+    // disk_capacity [-n|--node replica_server(ip:port)] [-o|--out file_name][-j|-json][-d|--detail]
+    const std::set<std::string> &params = {"n", "node", "o", "out"};
+    const std::set<std::string> &flags = {"j", "json", "d", "detail"};
+    argh::parser cmd(args.argc, args.argv, argh::parser::PREFER_PARAM_FOR_UNREG_OPTION);
+    if (!validate_cmd(cmd, params, flags)) {
+        return false;
+    }
+
+    bool format_to_json = cmd[{"-j", "--json"}];
+    bool query_detail_info = cmd[{"-d", "--detail"}];
+    std::string node_address = cmd({"-n", "--node"}).str();
+    std::string file_name = cmd({"-o", "--out"}).str();
+
+    std::ostream *ostream_ptr = get_out_stream(file_name);
+    if (!ostream_ptr) {
+        fmt::print(stderr, "get output stream failed!");
+        return false;
+    }
+    std::ostream &out = *ostream_ptr;
+
+    std::map<dsn::rpc_address, dsn::error_with<query_disk_info_response>> err_resps;
+    // passing empty app_name(app_name = "") means query all app disk info.
+    if (!query_disk_info(sc, cmd, node_address, "", err_resps)) {
+        if (ostream_ptr != &std::cout) {
```

Comment:
We'd better introduce something like `SCOPED_CLEANUP` in Kudu https://github.com/apache/kudu/blob/d0a93831ce39418c7d5c7d854a32aeb40b29dee7/src/kudu/util/scoped_cleanup.h#L33